### ISSUE TITLE
caching breaks implicit broadcasting

### ISSUE THREAD

**Issue by smartass101** [numba/numba#2120]
This bugs appears to be similar to #1466 because of the same error message with the mangled name, but is a different according to @pitrou, because this occurs only with caching on. Here is a failing test case with numba 0.28.1

```python
import numpy as np
import numba
import inspect


def test_implicit_broadcasting_with_caching():
    a = np.empty((2,3))
    def perform_smth(x):
        return x[0] * x[1]      # implicit broadcasting
    res_np = perform_smth(a)
    perf_numba = numba.jit(nopython=True, cache=False)(perform_smth)
    res_numba = perf_numba(a)
    np.testing.assert_allclose(res_np, res_numba) # passes fine without caching
    perf_numba_cache = numba.jit(nopython=True, cache=True)(perform_smth)
    try:
        res_numba_cache = perf_numba_cache(a) # try getting a cached version
    except KeyError as err:
        except_frame = inspect.trace()[-1][0]
        key_unmangled = '${}.{}'.format(*err.message.split('_')[1:])
        # the unmangled identifier is present, just $ and . are replaced with _
        assert key_unmangled in except_frame.f_locals['self'].typemap
        raise
    np.testing.assert_allclose(res_np, res_numba_cache) # the goal

```

**Comment by ma-ba** [comment]
Not reproducible with current master `0.29.0.dev+21.g6b17b3e`.
I got hit by this cache + name-mangling regression in `0.28.*` when using constants defined outside of a compiled function. The following is also **not** reproducible in `0.29.0.dev+21.g6b17b3e`:

```python
import numpy as np
from numba import njit

constant = 0
constant_array = np.zeros(1)

def f1(): return constant_array == constant_array
def f2(): return constant_array == 0
def f3(): return constant_array == np.zeros(1)
def f4(): return constant_array == constant
def f5(): return constant == np.zeros(1)

def compile_functions(*functions):
    for cache in (False, True):
        for func in functions:
            print("compiling", func.__name__, "with cache =", cache)
            try:
                njit(cache=cache)(func).compile(())
            except KeyError as e:
                # got KeyError: '_0_1' for all functions
                # with `cache=True` in Numba 0.28.*
                print(repr(e))
            else:
                print("success")

compile_functions(f1, f2, f3, f4, f5)
```

**Comment by pitrou** [comment]
Well, it appears I was wrong: they were really two instances of the same problem, or at least both were fixed by baf5e42ec64d9a6791f06d2150edbdbd6801c906. Thanks for posting the reproducers!
